src/io/g15.rs

```diff
@@ -0,0 +1,202 @@
+pub const REGEX_I_USERID: &str = r#"^m_iUserID\[(\d+)\]\s+integer\s+\((\d+)\)$"#;
+pub fn parse_userid(caps: Captures, players: &mut Vec<G15Player>) {
+    let idx = caps[1].parse::<usize>().unwrap();
+    let userid = caps[2].to_string();
+    players[idx].userid = userid;
+}
+
+// The G15Player struct contains all the data per player searched.
+#[derive(Debug, Clone)]
+pub struct G15Player {
+    name: String,     // eg "Lilith"
+    ping: u32,        // eg 21
+    score: u32,       // eg 16
+    deaths: u32,      // eg 5
+    sid3: String,     // eg [U:1:111216987]
+    team: u32,        // eg 3
+    health: u32,      // eg 125
+    ammo: u32,        // eg 6
+    connected: bool,  // eg true
+    valid: bool,      // eg true
+    alive: bool,      // eg true
+    userid: String,   // eg "301"
+}
+impl G15Player {
+    fn new() -> G15Player {
+        G15Player {
+            name: String::from("none"), 
+            ping: 0, 
+            score: 0, 
+            deaths: 0, 
+            sid3: String::from("none"), 
+            team: 0, 
+            health: 0, 
+            ammo: 0,
+            alive: false,
+            connected: false,
+            userid: String::from("0"),
+            valid: false,
+        }
+    }
+}
+
+/*
+  Parse a g15_dumpplayer string via Regex search.
+  We only extract useful data here, so drop most data.
+*/
+pub fn parse_g15(g15_log: String) -> Vec<G15Player> {
+    let mut players: Vec<G15Player> = vec![G15Player::new(); 34];
+    let patterns = [
+        RegMatch::new(REGEX_I_AMMO, parse_ammo), 
+        RegMatch::new(REGEX_SZ_NAME, parse_name),
+        RegMatch::new(REGEX_I_PING, parse_ping),
+        RegMatch::new(REGEX_I_SCORE, parse_score),
+        RegMatch::new(REGEX_I_DEATHS, parse_deaths),
+        RegMatch::new(REGEX_B_CONNECTED, parse_connected),
+        RegMatch::new(REGEX_I_TEAM, parse_team),
+        RegMatch::new(REGEX_B_ALIVE, parse_alive),
+        RegMatch::new(REGEX_I_HEALTH, parse_health),
+        RegMatch::new(REGEX_I_SID3, parse_sid3),
+        RegMatch::new(REGEX_B_VALID, parse_valid),
+        RegMatch::new(REGEX_I_USERID, parse_userid),
+        ];
+    let lines = g15_log.split("\n"); 
+
+    for line in lines {
+        for pat in &patterns {
+            // Perform a regex search on the current line using the stored regex in the current RegMatch struct
+            let pat_caps = pat.0.captures(line.trim());
+            if !pat_caps.is_none() {
+                // Call the corresponding function to store the capture groups, as we have found a match
+                pat.1(pat_caps.unwrap(), &mut players);
+            }
+        }
+    }
+    players.retain(|x| x.valid);
+
+    return players
+}
+
+
+fn main() {
+    println!("Reading test_g15.log file...");
+    let g15_log = fs::read_to_string("./test_g15.log")
+        .expect("No g15 log file found?");
+
+    let players = parse_g15(g15_log);
+
+    for pl in players {
+        println!("{:?}", pl);
+    }
+}
```

Comment:
This can be removed now, although makes me think we should probably have some tests for things like this.